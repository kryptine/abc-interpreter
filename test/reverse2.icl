module reverse2

// This is like the reverse example, but uses a different list: one with a Cons
// of arity 3. This is useful to test the garbage collector, because these
// nodes are stored differently.

import StdEnum
import StdInt
import StdList

:: List a = Cons a a (List a) | Nil

to_list :: [a] -> List a
to_list [x:xs] = Cons x x (to_list xs)
to_list []     = Nil

from_list :: (List a) -> [a]
from_list (Cons x _ xs) = [x:from_list xs]
from_list Nil           = []

NrOfTimes :== 1000

// Reversing a list of n integers n times.

MyReverse :: Int -> Int
MyReverse n = last (from_list (Rev_n n (to_list [1..n])))
where
	Rev_n :: Int (List a) -> List a
	Rev_n 1 list = Rev list Nil
	Rev_n n list = Rev_n (n - 1) (Rev list Nil)

	Rev :: (List a) (List a) -> List a
	Rev (Cons x y r) list = Rev r (Cons x y list)
	Rev Nil          list = list

Start :: Int
Start = MyReverse NrOfTimes
