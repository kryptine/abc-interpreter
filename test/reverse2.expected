1000
halt at 448
835719 1261433 2097152
